Features/4 yaml to json pipeline

radishmouse opened the pull request:
Closes #4 

This PR:
- adds dependencies to `requirements.txt`
- copies the `.yamllint` from https://github.com/thunderbird/thunderbird-notes/tree/master
- creates a `NotificationSchema` class that validates incoming yaml, produces equivalent json
- adds a python script that converts a folder of yaml files using `NotificationSchema`
  - uses `argparse` to accept input and output directory names
- adds a python file that generates a JSON schema file using `NotificationSchema` 
  - uses `argparse` to accept output file name

radishmouse commented on yaml-to-json.py:
```diff
@@ -0,0 +1,63 @@
+
+  def get_yaml_files(self):
+    # Using glob's ** pattern by specifying `recursive=True`
+    yaml_files = glob.glob(f'./{self.yaml_dir}/**/*.yaml', recursive=True)
```
Unsure if we want to search recursively, but adding it in here just in case.

radishmouse commented on yaml-to-json.py:
```diff
@@ -0,0 +1,63 @@
+
+  def yaml_to_json(self, in_file):
+    try:
+      return yaml.safe_load(in_file)
```
#3 mentions validation. I'm wondering if we could run a yaml linter as a step before the converter runs?

radishmouse commented on yaml-to-json.py:
```diff
@@ -0,0 +1,63 @@
+  def write_json(self, data_from_yaml, out_file):
+    try:
+      json.dump(data_from_yaml, out_file, indent=2)
+    except Exception as e:
```
Is it worth it to have a more specific error here?

radishmouse commented on yaml-to-json.py:
```diff
@@ -0,0 +1,63 @@
+    except Exception as e:
+      print(f'Error writing JSON file: {e}')
+
+  def convert(self):
```
This function felt kinda gross because of all the nesting, but I wasn't sure of a better way to structure it. Any suggestions?

radishmouse commented on yaml-to-json.py:
```diff
@@ -0,0 +1,63 @@
+              self.write_json(data_from_yaml, f)
+
+def main():
+  converter = YAMLtoJSONConverter(YAML_DIR, JSON_DIR)
+  converter.convert()
```
I'm assuming we wouldn't have a ton of notifications to process (or that it would take very long), but using a class could make it possible to split the yaml files into subdirectories and multiple converters could be run in parallel. 

Am I overthinking this? :nerd_face:

MelissaAutumn replied:
Possibly not worth it due to volume, but could be worth it by following a model schema. I would re-work this to load the yaml into [Pydantic](https://docs.pydantic.dev/latest/) for validation, and then you can json dump the model into a json file. 

(Static functions are defined by the lack of self as the first argument btw)

So:

``` python

class NotificationSchema(BaseModel):
  ... the schema goes here in python attributes ...

  def from_yaml(file_path):
    """Static function to generate NotificationSchema from a yaml file"""
    with open(file_path, 'r') as fh:
      contents = fh.read()
    ... read yaml to json and then to dictionary ...
    ... data = dictionary ...
    return NotificationSchema(**data)
      

in_path = '...'
out_path = '...'
files = get_file_list(from=in_path)

for file_name in files:
  schema = NotificationSchema.from_yaml(file_name)
  with open(join(out_path, file), 'w') as fh:
    fh.write(schema.model_dump_json())
```

Then if you'd like you can asyncio the inner loop.

MelissaAutumn reviewed (changes requested):
Look into Pydantic for data validation, it's like one of the best libraries ever and the reason why I enjoy Python https://docs.pydantic.dev/latest/

radishmouse commented:
@MelissaAutumn I might have gone a little too far for this PR, but I created the `NotificationSchema` using pydantic.

And I'm wondering: would this change (using pydantic to validate the incoming yaml and produce the json) basically close #1 and #3?

MelissaAutumn commented:
I'll look at this more in-depth tomorrow, but I believe you can generate a json schema file from a pydantic model https://docs.pydantic.dev/latest/concepts/json_schema/

Dump a copy, and check to see if it differs in any significant way from our current schema.

We can have a github action later on to auto-generate this per model change. (Manual is fine for now.)

radishmouse commented:
> I'll look at this more in-depth tomorrow, but I believe you can generate a json schema file from a pydantic model https://docs.pydantic.dev/latest/concepts/json_schema/
> 
> Dump a copy, and check to see if it differs in any significant way from our current schema.
> 
> We can have a github action later on to auto-generate this per model change. (Manual is fine for now.)

Looking at them side by side, the generated one is pretty close to the one submitted for #1.
The only thing not included in the generated schema is the "if CTA exists, URL must exist" requirement.

Also, the generated schema doesn't define min/max constraints for "percent_chance", so I'll add a `@field_validator` to account for that.

MelissaAutumn commented on NotificationSchema.py:
```diff
@@ -46,6 +46,13 @@ class Targeting(BaseModel):
 
+    @field_validator('percent_chance')
+    @classmethod
+    def min_max_percent_chance(cls, v: float):
```
I believe we can probably use https://docs.pydantic.dev/latest/concepts/fields/#numeric-constraints for this.

radishmouse replied:
ooh! nice. making that change now

radishmouse commented:
@MelissaAutumn I've made a few more changes:
- added the field validator for `percent_chance`
- created a schema generation method
- created a separate python script file for generating a JSON schema
- renamed the conversion script to `convert-yaml.py`

MelissaAutumn reviewed (commented):
Looking good, just a few more things! Feel free to punt some of these (like argparse) to a future ticket.

radishmouse commented on yaml-to-json.py:
```diff
@@ -0,0 +1,59 @@
+            f.write(json.dumps(schema, indent=2))
+
+def main():
+  converter = YAMLtoJSONConverter(YAML_DIR, JSON_DIR)
```
Can do!

radishmouse commented on yaml-to-json.py:
```diff
@@ -0,0 +1,59 @@
+
+  @classmethod
+  def generate_json_schema(cls):
+      schema = NotificationSchema.model_json_schema()  # (1)!
```
It is! I'm calling it in `generate-json-schema.py`

radishmouse commented:
@MelissaAutumn really awesome advice! I think it's ready for another pass.

MelissaAutumn reviewed (approved):
Looks good. Can you add a trailing new line at the end of each document. It's a python thing.
